dot/parachain/spam_slots.go

```diff
@@ -0,0 +1,200 @@
+}
+
+type slotKey struct {
+	session   parachain.SessionIndex
+	validator parachain.ValidatorIndex
+}
+
+type unconfirmedKey struct {
+	session   parachain.SessionIndex
+	candidate common.Hash
+}
+
+// spamSlots is an implementation of SpamSlots
+type spamSlots struct {
+	slots     map[slotKey]uint32
+	slotsLock sync.RWMutex
+
+	unconfirmed     map[unconfirmedKey]*treeset.Set
+	unconfirmedLock sync.RWMutex
+
+	maxSpamVotes uint32
+}
+
+// byValidatorIndex is a comparator for ValidatorIndex
+func byValidatorIndex(a, b interface{}) int {
+	return int(a.(parachain.ValidatorIndex) - b.(parachain.ValidatorIndex))
+}
+
+func slotsKey(session parachain.SessionIndex, validator parachain.ValidatorIndex) slotKey {
+	return slotKey{
+		session:   session,
+		validator: validator,
+	}
+}
+
+func unconfirmedDisputesKey(session parachain.SessionIndex, candidate common.Hash) unconfirmedKey {
+	return unconfirmedKey{
+		session:   session,
+		candidate: candidate,
+	}
+}
+
+func (s *spamSlots) getSpamCount(session parachain.SessionIndex, validator parachain.ValidatorIndex) (uint32, bool) {
+	s.slotsLock.RLock()
+	defer s.slotsLock.RUnlock()
+	spamCount, ok := s.slots[slotsKey(session, validator)]
+	return spamCount, ok
+}
+
+func (s *spamSlots) getValidators(session parachain.SessionIndex, candidate common.Hash) (*treeset.Set, bool) {
+	s.unconfirmedLock.RLock()
+	defer s.unconfirmedLock.RUnlock()
+	validators, ok := s.unconfirmed[unconfirmedDisputesKey(session, candidate)]
+	return validators, ok
+}
+
+func (s *spamSlots) clearValidators(session parachain.SessionIndex, candidate common.Hash) {
+	s.unconfirmedLock.Lock()
+	defer s.unconfirmedLock.Unlock()
+	delete(s.unconfirmed, unconfirmedDisputesKey(session, candidate))
+}
+
+func (s *spamSlots) clearSlots(session parachain.SessionIndex, validator parachain.ValidatorIndex) {
+	s.slotsLock.Lock()
+	defer s.slotsLock.Unlock()
+	delete(s.slots, slotsKey(session, validator))
+}
+
+func (s *spamSlots) incrementSpamCount(session parachain.SessionIndex, validator parachain.ValidatorIndex) {
+	s.slotsLock.Lock()
+	defer s.slotsLock.Unlock()
+	s.slots[slotsKey(session, validator)]++
+}
+
+func (s *spamSlots) AddUnconfirmed(session parachain.SessionIndex, candidate common.Hash, validator parachain.ValidatorIndex) bool {
+	if spamCount, _ := s.getSpamCount(session, validator); spamCount >= s.maxSpamVotes {
+		return false
+	}
+
+	s.unconfirmedLock.Lock()
+	defer s.unconfirmedLock.Unlock()
+
+	validators, ok := s.unconfirmed[unconfirmedDisputesKey(session, candidate)]
+	if !ok || validators == nil {
+		validators = treeset.NewWith(byValidatorIndex)
+	}
+
+	if !validators.Contains(validator) {
+		validators.Add(validator)
+		s.incrementSpamCount(session, validator)
+	}
+
+	s.unconfirmed[unconfirmedDisputesKey(session, candidate)] = validators
+
+	return true
+}
+
+func (s *spamSlots) Clear(session parachain.SessionIndex, candidate common.Hash) {
+	if validators, ok := s.getValidators(session, candidate); ok {
```

Comment:
I would suggest to use an early return avoiding nested logic
```suggestion
     validators, ok := s.getValidators(session, candidate)
     if !ok { return }
```